include/async/oneshot-event.hpp

```diff
@@ -24,4 +24,107 @@ struct oneshot_event {
 	wait_group wg_ { 1 };
 };
 
+struct oneshot_primitive {
+private:
+	struct node {
+		friend struct oneshot_primitive;
+
+		node(void (*complete)(node *))
+		: complete_{complete} { }
+
+		node(const node &) = delete;
+
+		node &operator= (const node &) = delete;
+
+	protected:
+		// Completion function.
+		// This saves a few nanoseconds in benchmarks compared to a virtual function.
+		void (*complete_)(node *);
```

Comment:
that seems odd, do you know why?

Reply:
There is one fewer indirection: `node->vtable->complete` vs. `node->complete` but no difference in object size (since the vtable would also take up one pointer).